stdlib/dynarray.mli

```diff
@@ -0,0 +1,641 @@
+(*                                                                        *)
+(**************************************************************************)
+
+(** Dynamic arrays.
+
+    The {!Array} module provide arrays of fixed length. {!Dynarray}
+    provides arrays whose length can change over time, by adding or
+    removing elements at the end of the array.
+
+    This is typically used to accumulate elements whose number is not
+    known in advance or changes during computation, while also
+    providing fast access to elements at arbitrary positions.
+
+{[
+    let dynarray_of_list li =
+      let arr = Dynarray.create () in
+      List.iter (fun v -> Dynarray.add_last arr v) li;
+      arr
+]}
+
+    The {!Buffer} module provides similar features, but it is
+    specialized for accumulating characters into a dynamically-resized
+    string.
+
+    The {!Stack} module provides a last-in first-out data structure
+    that can be easily implemented on top of dynamic arrays.
+
+    {b Warning.} In their current implementation, the memory layout
+    of dynamic arrays differs from the one of {!Array}s. See the
+    {{!section:memory_layout} Memory Layout} section for more information.
+
+    @since 5.2
+*)
+
+(** {b Unsynchronized accesses} *)
+
+[@@@alert unsynchronized_access
+    "Unsynchronized accesses to dynamic arrays are a programming error."
+]
+
+(**
+   Concurrent accesses to dynamic arrays must be synchronized
+   (for instance with a {!Mutex.t}). Unsynchronized accesses to
+   a dynamic array are a programming error that may lead to an invalid
+   dynamic array state, on which some operations would fail with an
+   [Invalid_argument] exception.
+*)
+
+(** {1:dynarrays Dynamic arrays} *)
+
+type 'a t
+(** A dynamic array containing values of type ['a].
+
+    A dynamic array [a] provides constant-time [get] and [set]
+    operations on indices between [0] and [Dynarray.length a - 1]
+    included. Its {!length} may change over time by adding or removing
+    elements to the end of the array.
+
+    We say that an index into a dynarray [a] is valid if it is in
+    [0 .. length a - 1] and invalid otherwise.
+*)
+
+val create : unit -> 'a t
+(** [create ()] is a new, empty array. *)
+
+val make : int -> 'a -> 'a t
+(** [make n x] is a new array of length [n], filled with [x].
+
+    @raise Invalid_argument if [n < 0] or [n > Sys.max_array_length].
+*)
+
+val init : int -> (int -> 'a) -> 'a t
+(** [init n f] is a new array [a] of length [n],
+    such that [get a i] is [f i]. In other words,
+    the elements of [a] are [f 0], then [f 1],
+    then [f 2]... and [f (n - 1)] last, evaluated
+    in that order.
+
+    This is similar to {!Array.init}.
+
+    @raise Invalid_argument if [n < 0] or [n > Sys.max_array_length].
+*)
+
+val get : 'a t -> int -> 'a
+(** [get a i] is the [i]-th element of [a], starting with index [0].
+
+    @raise Invalid_argument if the index is invalid *)
```

Comment:
Same as above the `Invalid_argument` leaks information about the implementation but that's is probably unavoidable in term of performance here.

Reply:
Yes, this is a case where there is a noticeable performance benefit to do so, for a critical function. This choice is documented by an implementation comment. (In all other cases that are not performance-critical, I reverted this use of `Array.get` on suggestion of @gadmm, preferring clear errors to a slightly more compact bound check.)